src/target/riscv/riscv.c

```diff
@@ -5365,14 +5318,39 @@ static struct reg_arch_type riscv_reg_arch_type = {
 	.set = register_set
 };
 
-struct csr_info {
-	unsigned number;
-	const char *name;
-};
-
-static int cmp_csr_info(const void *p1, const void *p2)
+static int initialize_custom_register_names(struct list_head *expose_custom,
+		struct reg_name_table *custom_register_names)
 {
-	return (int) (((struct csr_info *)p1)->number) - (int) (((struct csr_info *)p2)->number);
+	unsigned int custom_regs_num = 0;
+	if (!list_empty(expose_custom)) {
+		range_list_t *entry;
+		list_for_each_entry(entry, expose_custom, list)
+			custom_regs_num += entry->high - entry->low + 1;
+	}
+
+	if (!custom_regs_num)
+		return ERROR_OK;
+
+	custom_register_names->reg_names = calloc(custom_regs_num, sizeof(struct reg_name));
+	custom_register_names->names_num = custom_regs_num;
+	if (!custom_register_names->reg_names) {
+		LOG_ERROR("Failed to allocate memory for custom_register_names->reg_names");
+		return ERROR_FAIL;
+	}
+	struct reg_name *additional_regs = custom_register_names->reg_names;
+	range_list_t *range;
+	unsigned int custom_registers_index = 0;
+	list_for_each_entry(range, expose_custom, list) {
+		for (unsigned int custom_number = range->low; custom_number <= range->high; ++custom_number) {
+			if (range->name)
+				initialize_reg_name(&additional_regs[custom_registers_index], range->name);
```

Comment:
It would be better to point directly at range->name. That way you're not limiting the user to any register length.

It might be simpler to just malloc() each register. A bit wasteful, but less wasteful than statically allocating 32 bytes, and then changing the pointer is trivial. Needs slightly more free() code, but should also be straightforward.

Reply:
In this case, the memory consumption is about 128 kilobytes, which is very small for current devices. I tried to keep the old behavior and not add dynamic memory, so as not to create potential memory bugs. If you insist I can change it.

Reply:
I really do prefer one malloc() per register. It will remove a limit on the register size a user can type, and the code complexity won't change much. I also think that version of the code will be easier to upstream.